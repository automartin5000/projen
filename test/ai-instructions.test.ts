import { synthSnapshot, TestProject } from "./util";
import { AiInstructions, AiAgent } from "../src/ai-instructions";

describe("AiInstructions", () => {
  test("generates files for all AI agents by default", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot[".github/copilot-instructions.md"]).toBeDefined();
    expect(snapshot[".cursor/rules/projen.md"]).toBeDefined();
    expect(snapshot["CLAUDE.md"]).toBeDefined();
    expect(snapshot[".amazonq/rules/projen.md"]).toBeDefined();
  });

  test("default instructions include projen-specific guidance", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const copilotInstructions = snapshot[".github/copilot-instructions.md"];

    expect(copilotInstructions).toContain("projen");
    expect(copilotInstructions).toContain("npx projen");
    expect(copilotInstructions).toContain(
      "DO NOT manually edit generated files"
    );
    expect(copilotInstructions).toContain("Modify configuration in .projenrc");
  });

  test("can specify which AI agents to support", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project, {
      supportedAiAgents: [AiAgent.GITHUB_COPILOT, AiAgent.CURSOR],
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot[".github/copilot-instructions.md"]).toBeDefined();
    expect(snapshot[".cursor/rules/projen.md"]).toBeDefined();
    expect(snapshot["CLAUDE.md"]).toBeUndefined();
    expect(snapshot[".amazonq/rules/projen.md"]).toBeUndefined();
  });

  test("can specify only one AI agent", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project, {
      supportedAiAgents: [AiAgent.GITHUB_COPILOT],
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot[".github/copilot-instructions.md"]).toBeDefined();
    expect(snapshot[".cursor/rules/projen.md"]).toBeUndefined();
    expect(snapshot["CLAUDE.md"]).toBeUndefined();
    expect(snapshot[".amazonq/rules/projen.md"]).toBeUndefined();
  });

  test("extracts task runner from projenCommand - npx", () => {
    // GIVEN
    const project = new TestProject({
      projenCommand: "npx projen",
    });

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("npx projen");
    expect(instructions).not.toContain("bunx projen");
  });

  test("extracts task runner from projenCommand - bunx", () => {
    // GIVEN
    const project = new TestProject({
      projenCommand: "bunx projen",
    });

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("bunx projen");
    expect(instructions).not.toContain("npx projen");
  });

  test("extracts task runner from projenCommand - pnpm", () => {
    // GIVEN
    const project = new TestProject({
      projenCommand: "pnpm projen",
    });

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("pnpm projen");
  });

  test("extracts task runner from projenCommand - yarn", () => {
    // GIVEN
    const project = new TestProject({
      projenCommand: "yarn projen",
    });

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("yarn projen");
  });

  test("addCustomInstruction adds general instructions", () => {
    // GIVEN
    const project = new TestProject();
    const ai = new AiInstructions(project);

    // WHEN
    ai.addCustomInstruction("Always use functional programming patterns.");
    ai.addCustomInstruction("Prefer immutability.");

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("Project-Specific Instructions");
    expect(instructions).toContain(
      "Always use functional programming patterns."
    );
    expect(instructions).toContain("Prefer immutability.");
  });

  test("addAgentInstruction adds agent-specific instructions", () => {
    // GIVEN
    const project = new TestProject();
    const ai = new AiInstructions(project);

    // WHEN
    ai.addAgentInstruction(
      AiAgent.GITHUB_COPILOT,
      "Use descriptive commit messages."
    );

    // THEN
    const snapshot = synthSnapshot(project);
    const copilotInstructions = snapshot[".github/copilot-instructions.md"];
    const cursorInstructions = snapshot[".cursor/rules/projen.md"];

    expect(copilotInstructions).toContain(
      "GitHub Copilot-Specific Instructions"
    );
    expect(copilotInstructions).toContain("Use descriptive commit messages.");
    expect(cursorInstructions).not.toContain(
      "GitHub Copilot-Specific Instructions"
    );
  });

  test("perAgentInstructions option with string", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project, {
      perAgentInstructions: {
        [AiAgent.CURSOR]: {
          instructions: ["Cursor-specific instruction."],
        },
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    const cursorInstructions = snapshot[".cursor/rules/projen.md"];

    expect(cursorInstructions).toContain("Cursor-Specific Instructions");
    expect(cursorInstructions).toContain("Cursor-specific instruction.");
  });

  test("perAgentInstructions option with array", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project, {
      perAgentInstructions: {
        [AiAgent.CLAUDE]: {
          instructions: ["First instruction.", "Second instruction."],
        },
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    const claudeInstructions = snapshot["CLAUDE.md"];

    expect(claudeInstructions).toContain("Claude-Specific Instructions");
    expect(claudeInstructions).toContain("First instruction.");
    expect(claudeInstructions).toContain("Second instruction.");
  });

  test("generated files are marked as readonly", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    const copilotFile = project.tryFindFile(".github/copilot-instructions.md");
    const cursorFile = project.tryFindFile(".cursor/rules/projen.md");
    const claudeFile = project.tryFindFile("CLAUDE.md");
    const amazonQFile = project.tryFindFile(".amazonq/rules/projen.md");

    expect(copilotFile?.readonly).toBe(true);
    expect(cursorFile?.readonly).toBe(true);
    expect(claudeFile?.readonly).toBe(true);
    expect(amazonQFile?.readonly).toBe(true);
  });

  test("all agent files have consistent base content", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const copilotContent = snapshot[".github/copilot-instructions.md"];
    const cursorContent = snapshot[".cursor/rules/projen.md"];
    const claudeContent = snapshot["CLAUDE.md"];
    const amazonQContent = snapshot[".amazonq/rules/projen.md"];

    // All should contain base projen instructions
    expect(copilotContent).toContain("Projen Project Instructions");
    expect(cursorContent).toContain("Projen Project Instructions");
    expect(claudeContent).toContain("Projen Project Instructions");
    expect(amazonQContent).toContain("Projen Project Instructions");
  });

  test("includes common projen tasks in instructions", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("build");
    expect(instructions).toContain("test");
    expect(instructions).toContain("compile");
    expect(instructions).toContain(".projen/tasks.json");
  });

  test("includes dependency management guidance", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("Add dependencies through projen");
    expect(instructions).toContain("addDeps()");
    expect(instructions).toContain("addDevDeps()");
  });

  test("includes workflow guidance", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("Workflow");
    expect(instructions).toContain("Make changes to .projenrc");
    expect(instructions).toContain("synthesize");
  });

  test("combining general and agent-specific instructions", () => {
    // GIVEN
    const project = new TestProject();
    const ai = new AiInstructions(project, {
      supportedAiAgents: [AiAgent.GITHUB_COPILOT, AiAgent.CURSOR],
      perAgentInstructions: {
        [AiAgent.GITHUB_COPILOT]: {
          instructions: ["Copilot-specific rule."],
        },
      },
    });

    // WHEN
    ai.addCustomInstruction("General instruction for all agents.");

    // THEN
    const snapshot = synthSnapshot(project);
    const copilotInstructions = snapshot[".github/copilot-instructions.md"];
    const cursorInstructions = snapshot[".cursor/rules/projen.md"];

    // Both should have general instruction
    expect(copilotInstructions).toContain(
      "General instruction for all agents."
    );
    expect(cursorInstructions).toContain("General instruction for all agents.");

    // Only copilot should have copilot-specific
    expect(copilotInstructions).toContain("Copilot-specific rule.");
    expect(cursorInstructions).not.toContain("Copilot-specific rule.");
  });

  test("snapshots - default configuration", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    expect(synthSnapshot(project)).toMatchSnapshot();
  });

  test("snapshots - custom agents and instructions", () => {
    // GIVEN
    const project = new TestProject({
      projenCommand: "bunx projen",
    });
    const ai = new AiInstructions(project, {
      supportedAiAgents: [AiAgent.GITHUB_COPILOT, AiAgent.CURSOR],
      perAgentInstructions: {
        [AiAgent.CURSOR]: {
          instructions: ["Use TypeScript strict mode.", "Always write tests."],
        },
      },
    });

    // WHEN
    ai.addCustomInstruction("Follow the project's coding standards.");

    // THEN
    expect(synthSnapshot(project)).toMatchSnapshot();
  });

  test("includes build instructions by default", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project);

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("Development Best Practices");
    expect(instructions).toContain("Always run build after changes");
    expect(instructions).toContain("Task completion criteria");
    expect(instructions).toContain("All tests pass");
    expect(instructions).toContain("no compilation errors");
    expect(instructions).toContain("no linting errors");
  });

  test("can disable build instructions", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    new AiInstructions(project, {
      includeBuildInstructions: false,
    });

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).not.toContain("Development Best Practices");
    expect(instructions).not.toContain("Always run build after changes");
  });

  test("supports variadic addCustomInstruction", () => {
    // GIVEN
    const project = new TestProject();
    const ai = new AiInstructions(project);

    // WHEN
    ai.addCustomInstruction(
      "Use functional programming patterns.",
      "Prefer immutability.",
      "Always write tests."
    );

    // THEN
    const snapshot = synthSnapshot(project);
    const instructions = snapshot[".github/copilot-instructions.md"];

    expect(instructions).toContain("Use functional programming patterns.");
    expect(instructions).toContain("Prefer immutability.");
    expect(instructions).toContain("Always write tests.");
  });

  test("supports variadic addAgentInstruction", () => {
    // GIVEN
    const project = new TestProject();
    const ai = new AiInstructions(project);

    // WHEN
    ai.addAgentInstruction(
      AiAgent.GITHUB_COPILOT,
      "Use descriptive commit messages.",
      "Keep PRs small and focused.",
      "Write clear documentation."
    );

    // THEN
    const snapshot = synthSnapshot(project);
    const copilotInstructions = snapshot[".github/copilot-instructions.md"];

    expect(copilotInstructions).toContain("Use descriptive commit messages.");
    expect(copilotInstructions).toContain("Keep PRs small and focused.");
    expect(copilotInstructions).toContain("Write clear documentation.");
  });
});
