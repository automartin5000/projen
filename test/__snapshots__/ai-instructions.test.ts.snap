// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AiInstructions snapshots - custom agents and instructions 1`] = `
{
  ".cursor/rules/projen.md": "# Projen Project Instructions

This project is managed by [projen](https://github.com/projen/projen), a project configuration management tool.

## Important Guidelines

### Task Execution

- **Always use projen for task execution**: Run tasks using \`bunx projen <task-name>\` instead of directly using npm, yarn, or other package managers.
- **Check available tasks**: Look in \`.projen/tasks.json\` to see all available tasks, their descriptions, and steps.
- **Common tasks**: 
  - \`bunx projen\` - Synthesize project configuration files
  - \`bunx projen build\` - Build the project
  - \`bunx projen test\` - Run tests
  - \`bunx projen compile\` - Compile the source code

### File Modifications

- **DO NOT manually edit generated files**: Files marked with a comment like "~~ Generated by projen. To modify, edit .projenrc.js and run \`bunx projen\`" should never be edited directly.
- **Modify configuration in .projenrc**: To change project configuration, always edit the \`.projenrc.js\`, \`.projenrc.ts\`, or \`.projenrc.json\` file and then run \`bunx projen\` to regenerate the project files.
- **Check .projenrc first**: Before suggesting changes to package.json, tsconfig.json, or other configuration files, always check if these are managed by projen and suggest changes to .projenrc instead.

### Dependencies

- **Add dependencies through projen**: Use the projen configuration to add dependencies instead of manually editing package.json or using npm/yarn install directly.
- **Example**: In .projenrc.js, use methods like \`addDeps()\`, \`addDevDeps()\`, or \`addPeerDeps()\` to add dependencies.


### Development Best Practices

- **Always run build after changes**: After modifying any source or test file, run \`bunx projen build\` to ensure your changes compile and pass all tests.
- **Task completion criteria**: A task is not considered complete until:
  - All tests pass (\`bunx projen test\`)
  - There are no compilation errors (\`bunx projen compile\`)
  - There are no linting errors (usually part of the build, if not, run the linter defined in tasks.json)
  - The full build succeeds (\`bunx projen build\`)

### Workflow

1. Make changes to .projenrc configuration file
2. Run \`bunx projen\` to synthesize and update generated files
3. Review the changes
4. Commit both .projenrc and the generated files

## Projen Configuration

This project's configuration is defined in the .projenrc file at the root of the repository. All project metadata, dependencies, scripts, and tooling configuration should be managed through this file.

## Additional Resources

- [Projen Documentation](https://projen.io)
- [Projen GitHub Repository](https://github.com/projen/projen)
## Project-Specific Instructions

Follow the project's coding standards.
## Cursor-Specific Instructions

Use TypeScript strict mode.

Always write tests.",
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "bunx projen".

* text=auto eol=lf
/.cursor/rules/projen.md linguist-generated
/.gitattributes linguist-generated
/.github/copilot-instructions.md linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/copilot-instructions.md": "# Projen Project Instructions

This project is managed by [projen](https://github.com/projen/projen), a project configuration management tool.

## Important Guidelines

### Task Execution

- **Always use projen for task execution**: Run tasks using \`bunx projen <task-name>\` instead of directly using npm, yarn, or other package managers.
- **Check available tasks**: Look in \`.projen/tasks.json\` to see all available tasks, their descriptions, and steps.
- **Common tasks**: 
  - \`bunx projen\` - Synthesize project configuration files
  - \`bunx projen build\` - Build the project
  - \`bunx projen test\` - Run tests
  - \`bunx projen compile\` - Compile the source code

### File Modifications

- **DO NOT manually edit generated files**: Files marked with a comment like "~~ Generated by projen. To modify, edit .projenrc.js and run \`bunx projen\`" should never be edited directly.
- **Modify configuration in .projenrc**: To change project configuration, always edit the \`.projenrc.js\`, \`.projenrc.ts\`, or \`.projenrc.json\` file and then run \`bunx projen\` to regenerate the project files.
- **Check .projenrc first**: Before suggesting changes to package.json, tsconfig.json, or other configuration files, always check if these are managed by projen and suggest changes to .projenrc instead.

### Dependencies

- **Add dependencies through projen**: Use the projen configuration to add dependencies instead of manually editing package.json or using npm/yarn install directly.
- **Example**: In .projenrc.js, use methods like \`addDeps()\`, \`addDevDeps()\`, or \`addPeerDeps()\` to add dependencies.


### Development Best Practices

- **Always run build after changes**: After modifying any source or test file, run \`bunx projen build\` to ensure your changes compile and pass all tests.
- **Task completion criteria**: A task is not considered complete until:
  - All tests pass (\`bunx projen test\`)
  - There are no compilation errors (\`bunx projen compile\`)
  - There are no linting errors (usually part of the build, if not, run the linter defined in tasks.json)
  - The full build succeeds (\`bunx projen build\`)

### Workflow

1. Make changes to .projenrc configuration file
2. Run \`bunx projen\` to synthesize and update generated files
3. Review the changes
4. Commit both .projenrc and the generated files

## Projen Configuration

This project's configuration is defined in the .projenrc file at the root of the repository. All project metadata, dependencies, scripts, and tooling configuration should be managed through this file.

## Additional Resources

- [Projen Documentation](https://projen.io)
- [Projen GitHub Repository](https://github.com/projen/projen)
## Project-Specific Instructions

Follow the project's coding standards.",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "bunx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "bunx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.github/copilot-instructions.md
!/.cursor/rules/projen.md
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "bunx projen".",
    "files": [
      ".cursor/rules/projen.md",
      ".gitattributes",
      ".github/copilot-instructions.md",
      ".github/workflows/pull-request-lint.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "bunx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "README.md": "# replace this",
}
`;

exports[`AiInstructions snapshots - default configuration 1`] = `
{
  ".amazonq/rules/projen.md": "# Projen Project Instructions

This project is managed by [projen](https://github.com/projen/projen), a project configuration management tool.

## Important Guidelines

### Task Execution

- **Always use projen for task execution**: Run tasks using \`npx projen <task-name>\` instead of directly using npm, yarn, or other package managers.
- **Check available tasks**: Look in \`.projen/tasks.json\` to see all available tasks, their descriptions, and steps.
- **Common tasks**: 
  - \`npx projen\` - Synthesize project configuration files
  - \`npx projen build\` - Build the project
  - \`npx projen test\` - Run tests
  - \`npx projen compile\` - Compile the source code

### File Modifications

- **DO NOT manually edit generated files**: Files marked with a comment like "~~ Generated by projen. To modify, edit .projenrc.js and run \`npx projen\`" should never be edited directly.
- **Modify configuration in .projenrc**: To change project configuration, always edit the \`.projenrc.js\`, \`.projenrc.ts\`, or \`.projenrc.json\` file and then run \`npx projen\` to regenerate the project files.
- **Check .projenrc first**: Before suggesting changes to package.json, tsconfig.json, or other configuration files, always check if these are managed by projen and suggest changes to .projenrc instead.

### Dependencies

- **Add dependencies through projen**: Use the projen configuration to add dependencies instead of manually editing package.json or using npm/yarn install directly.
- **Example**: In .projenrc.js, use methods like \`addDeps()\`, \`addDevDeps()\`, or \`addPeerDeps()\` to add dependencies.


### Development Best Practices

- **Always run build after changes**: After modifying any source or test file, run \`npx projen build\` to ensure your changes compile and pass all tests.
- **Task completion criteria**: A task is not considered complete until:
  - All tests pass (\`npx projen test\`)
  - There are no compilation errors (\`npx projen compile\`)
  - There are no linting errors (usually part of the build, if not, run the linter defined in tasks.json)
  - The full build succeeds (\`npx projen build\`)

### Workflow

1. Make changes to .projenrc configuration file
2. Run \`npx projen\` to synthesize and update generated files
3. Review the changes
4. Commit both .projenrc and the generated files

## Projen Configuration

This project's configuration is defined in the .projenrc file at the root of the repository. All project metadata, dependencies, scripts, and tooling configuration should be managed through this file.

## Additional Resources

- [Projen Documentation](https://projen.io)
- [Projen GitHub Repository](https://github.com/projen/projen)",
  ".cursor/rules/projen.md": "# Projen Project Instructions

This project is managed by [projen](https://github.com/projen/projen), a project configuration management tool.

## Important Guidelines

### Task Execution

- **Always use projen for task execution**: Run tasks using \`npx projen <task-name>\` instead of directly using npm, yarn, or other package managers.
- **Check available tasks**: Look in \`.projen/tasks.json\` to see all available tasks, their descriptions, and steps.
- **Common tasks**: 
  - \`npx projen\` - Synthesize project configuration files
  - \`npx projen build\` - Build the project
  - \`npx projen test\` - Run tests
  - \`npx projen compile\` - Compile the source code

### File Modifications

- **DO NOT manually edit generated files**: Files marked with a comment like "~~ Generated by projen. To modify, edit .projenrc.js and run \`npx projen\`" should never be edited directly.
- **Modify configuration in .projenrc**: To change project configuration, always edit the \`.projenrc.js\`, \`.projenrc.ts\`, or \`.projenrc.json\` file and then run \`npx projen\` to regenerate the project files.
- **Check .projenrc first**: Before suggesting changes to package.json, tsconfig.json, or other configuration files, always check if these are managed by projen and suggest changes to .projenrc instead.

### Dependencies

- **Add dependencies through projen**: Use the projen configuration to add dependencies instead of manually editing package.json or using npm/yarn install directly.
- **Example**: In .projenrc.js, use methods like \`addDeps()\`, \`addDevDeps()\`, or \`addPeerDeps()\` to add dependencies.


### Development Best Practices

- **Always run build after changes**: After modifying any source or test file, run \`npx projen build\` to ensure your changes compile and pass all tests.
- **Task completion criteria**: A task is not considered complete until:
  - All tests pass (\`npx projen test\`)
  - There are no compilation errors (\`npx projen compile\`)
  - There are no linting errors (usually part of the build, if not, run the linter defined in tasks.json)
  - The full build succeeds (\`npx projen build\`)

### Workflow

1. Make changes to .projenrc configuration file
2. Run \`npx projen\` to synthesize and update generated files
3. Review the changes
4. Commit both .projenrc and the generated files

## Projen Configuration

This project's configuration is defined in the .projenrc file at the root of the repository. All project metadata, dependencies, scripts, and tooling configuration should be managed through this file.

## Additional Resources

- [Projen Documentation](https://projen.io)
- [Projen GitHub Repository](https://github.com/projen/projen)",
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf
/.amazonq/rules/projen.md linguist-generated
/.cursor/rules/projen.md linguist-generated
/.gitattributes linguist-generated
/.github/copilot-instructions.md linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated
/CLAUDE.md linguist-generated",
  ".github/copilot-instructions.md": "# Projen Project Instructions

This project is managed by [projen](https://github.com/projen/projen), a project configuration management tool.

## Important Guidelines

### Task Execution

- **Always use projen for task execution**: Run tasks using \`npx projen <task-name>\` instead of directly using npm, yarn, or other package managers.
- **Check available tasks**: Look in \`.projen/tasks.json\` to see all available tasks, their descriptions, and steps.
- **Common tasks**: 
  - \`npx projen\` - Synthesize project configuration files
  - \`npx projen build\` - Build the project
  - \`npx projen test\` - Run tests
  - \`npx projen compile\` - Compile the source code

### File Modifications

- **DO NOT manually edit generated files**: Files marked with a comment like "~~ Generated by projen. To modify, edit .projenrc.js and run \`npx projen\`" should never be edited directly.
- **Modify configuration in .projenrc**: To change project configuration, always edit the \`.projenrc.js\`, \`.projenrc.ts\`, or \`.projenrc.json\` file and then run \`npx projen\` to regenerate the project files.
- **Check .projenrc first**: Before suggesting changes to package.json, tsconfig.json, or other configuration files, always check if these are managed by projen and suggest changes to .projenrc instead.

### Dependencies

- **Add dependencies through projen**: Use the projen configuration to add dependencies instead of manually editing package.json or using npm/yarn install directly.
- **Example**: In .projenrc.js, use methods like \`addDeps()\`, \`addDevDeps()\`, or \`addPeerDeps()\` to add dependencies.


### Development Best Practices

- **Always run build after changes**: After modifying any source or test file, run \`npx projen build\` to ensure your changes compile and pass all tests.
- **Task completion criteria**: A task is not considered complete until:
  - All tests pass (\`npx projen test\`)
  - There are no compilation errors (\`npx projen compile\`)
  - There are no linting errors (usually part of the build, if not, run the linter defined in tasks.json)
  - The full build succeeds (\`npx projen build\`)

### Workflow

1. Make changes to .projenrc configuration file
2. Run \`npx projen\` to synthesize and update generated files
3. Review the changes
4. Commit both .projenrc and the generated files

## Projen Configuration

This project's configuration is defined in the .projenrc file at the root of the repository. All project metadata, dependencies, scripts, and tooling configuration should be managed through this file.

## Additional Resources

- [Projen Documentation](https://projen.io)
- [Projen GitHub Repository](https://github.com/projen/projen)",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.github/copilot-instructions.md
!/.cursor/rules/projen.md
!/CLAUDE.md
!/.amazonq/rules/projen.md
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".amazonq/rules/projen.md",
      ".cursor/rules/projen.md",
      ".gitattributes",
      ".github/copilot-instructions.md",
      ".github/workflows/pull-request-lint.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "CLAUDE.md",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "CLAUDE.md": "# Projen Project Instructions

This project is managed by [projen](https://github.com/projen/projen), a project configuration management tool.

## Important Guidelines

### Task Execution

- **Always use projen for task execution**: Run tasks using \`npx projen <task-name>\` instead of directly using npm, yarn, or other package managers.
- **Check available tasks**: Look in \`.projen/tasks.json\` to see all available tasks, their descriptions, and steps.
- **Common tasks**: 
  - \`npx projen\` - Synthesize project configuration files
  - \`npx projen build\` - Build the project
  - \`npx projen test\` - Run tests
  - \`npx projen compile\` - Compile the source code

### File Modifications

- **DO NOT manually edit generated files**: Files marked with a comment like "~~ Generated by projen. To modify, edit .projenrc.js and run \`npx projen\`" should never be edited directly.
- **Modify configuration in .projenrc**: To change project configuration, always edit the \`.projenrc.js\`, \`.projenrc.ts\`, or \`.projenrc.json\` file and then run \`npx projen\` to regenerate the project files.
- **Check .projenrc first**: Before suggesting changes to package.json, tsconfig.json, or other configuration files, always check if these are managed by projen and suggest changes to .projenrc instead.

### Dependencies

- **Add dependencies through projen**: Use the projen configuration to add dependencies instead of manually editing package.json or using npm/yarn install directly.
- **Example**: In .projenrc.js, use methods like \`addDeps()\`, \`addDevDeps()\`, or \`addPeerDeps()\` to add dependencies.


### Development Best Practices

- **Always run build after changes**: After modifying any source or test file, run \`npx projen build\` to ensure your changes compile and pass all tests.
- **Task completion criteria**: A task is not considered complete until:
  - All tests pass (\`npx projen test\`)
  - There are no compilation errors (\`npx projen compile\`)
  - There are no linting errors (usually part of the build, if not, run the linter defined in tasks.json)
  - The full build succeeds (\`npx projen build\`)

### Workflow

1. Make changes to .projenrc configuration file
2. Run \`npx projen\` to synthesize and update generated files
3. Review the changes
4. Commit both .projenrc and the generated files

## Projen Configuration

This project's configuration is defined in the .projenrc file at the root of the repository. All project metadata, dependencies, scripts, and tooling configuration should be managed through this file.

## Additional Resources

- [Projen Documentation](https://projen.io)
- [Projen GitHub Repository](https://github.com/projen/projen)",
  "README.md": "# replace this",
}
`;
